akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala

```diff
@@ -0,0 +1,434 @@
+   *
+   *            +---------------------------------------+-----------------+
+   * inOwners:  | index to stages array                 | Boundary (-1)   |
+   *            +----------------+----------------------+-----------------+
+   * ins:       | exposed inputs | internal connections | nulls           |
+   *            +----------------+----------------------+-----------------+
+   * outs:      | nulls          | internal connections | exposed outputs |
+   *            +----------------+----------------------+-----------------+
+   * outOwners: | Boundary (-1)  | index to stages array                  |
+   *            +----------------+----------------------------------------+
+   *
+   * In addition, it is also assumed by the infrastructure that the order of exposed inputs and outputs in the
+   * corresponding segments of these arrays matches the exact same order of the ports in the [[Shape]].
+   *
+   */
+  final case class GraphAssembly(stages: Array[GraphStage[_]],
+                                 ins: Array[Inlet[_]],
+                                 inOwners: Array[Int],
+                                 outs: Array[Outlet[_]],
+                                 outOwners: Array[Int]) {
+
+    val connectionCount: Int = ins.length
+
+    /**
+     * Takes an interpreter and returns three arrays required by the interpreter containing the input, output port
+     * handlers and the stage logic instances.
+     */
+    def materialize(interpreter: GraphInterpreter): (Array[InHandler], Array[OutHandler], Array[GraphStageLogic]) = {
+      val logics = Array.ofDim[GraphStageLogic](stages.length)
+      for (i ← stages.indices) {
+        logics(i) = stages(i).createLogic
+        logics(i).interpreter = interpreter
+      }
+
+      val inHandlers = Array.ofDim[InHandler](connectionCount)
+      val outHandlers = Array.ofDim[OutHandler](connectionCount)
+
+      for (i ← 0 until connectionCount) {
+        if (ins(i) ne null) {
+          inHandlers(i) = logics(inOwners(i)).inHandlers(ins(i))
+          logics(inOwners(i)).inToConn += ins(i) -> i
+        }
+        if (outs(i) ne null) {
+          outHandlers(i) = logics(outOwners(i)).outHandlers(outs(i))
+          logics(outOwners(i)).outToConn += outs(i) -> i
+        }
+      }
+
+      (inHandlers, outHandlers, logics)
+    }
+
+    override def toString: String =
+      "GraphAssembly(" +
+        stages.mkString("[", ",", "]") + ", " +
+        ins.mkString("[", ",", "]") + ", " +
+        inOwners.mkString("[", ",", "]") + ", " +
+        outs.mkString("[", ",", "]") + ", " +
+        outOwners.mkString("[", ",", "]") +
+        ")"
+  }
+}
+
+/**
+ * INERNAL API
+ *
+ * From an external viewpoint, the GraphInterpreter takes an assembly of graph processing stages encoded as a
+ * [[GraphInterpreter#GraphAssembly]] object and provides facilities to execute and interact with this assembly.
```

Comment:
@drewhk this is one of the special cases that doesn't work, nested object I guess